oeps/oep-0035-bp-rest-api-pagination.rst

```diff
@@ -0,0 +1,257 @@
+
+Expensive Counts
+++++++++++++++++
+
+Returning the ``count()`` of all records that could be paged through becomes
+expensive in and of itself. Counting millions of records can take seconds just
+by itself. This is true even if the current page you're returning only has ten
+results in it.
+
+Fish Tailing
+++++++++++++
+
+You will often see APIs where requests to pages get more and more expensive as
+you traverse the result set. When doing numbered pages, the database typically
+does a query like ``LIMIT {page_size} OFFSET {records to start of current
+page}``. But each time it does that, the database has to examine all the records
+leading up to the ones being displayed. Meaning that if you show 100 records
+starting at an offset of 1,000,000, then the database has to examine 1,000,100
+rows.
+
+Incorrect Indexes
++++++++++++++++++
+
+As APIs become more complicated with nested relationships and sorting options,
+it's possible to accidentally introduce queries that do some combination of
+table scans and file sorting, either one of which will cripple performance and
+threaten system stability.
+
+Too Much Prefetch-Related
++++++++++++++++++++++++++
+
+Django's `prefetch_related` is an important tool for avoiding "n+1" style
+queries, and is the right thing to do most of the time. However, highly
+normalized schemas with many nested relations in a single view may cause CPU
+issues because of how `prefetch_related` is implemented by Django. This has been
+an issue with the course discovery service.
+
+
+Recommendations by Use Case
+---------------------------
+
+This OEP is going to examine pagination over three common listing types.
+
+Use Case 1: Unbounded Lists of Resources for Human Interaction
+==============================================================
+
+This is the most common case for pagination, where the pages returned by an API
+correspond to the UI displayed to an end user's browser. The total, unfiltered
+listing of resources may be millions or billions of rows from the database, but
+the user only needs to see a small fraction of that.
+
+Recommendations
++++++++++++++++
+
+Use Django Rest Framework's default ``PageNumberPagination`` class. It will give
+the following general template::
+
+  {
+    "count": 1023
+    "next": "https://api.example.org/accounts/?page=5",
+    "previous": "https://api.example.org/accounts/?page=3",
+    "results": [
+       …
+    ]
+  }
+
+The query parameter for the page number is ``page``.
+
+Your API should accept a ``page_size`` parameter, by defining
+``page_size_query_param``. However if you allow clients to specify the desired
+page size, your *must* specify a ``max_page_size``.
+
+Finally, you *must* limit the number of items in your result set. You can do
+this by modifying your view's ``queryset``. For example::
+
+  class UserViewSet(viewsets.ModelViewSet):
+      serializer_class = UserSerializer
+      queryset = User.objects.all().limit(1000)  # <-- Limit is here
```

Comment:
@ormsbee do you have a working DRF example of limiting the result set that I can check out?

Django querysets do not have a `.limit(...)` method, and using a lazy slice limit (`queryset = User.objects.all()[:1000]`) does not work because Django "[cannot](https://stackoverflow.com/a/30020676) filter a query once a slice has been taken." and DRF needs to filter the queryset later on.

The DRF paginator could be modified to have a `max_page` argument, but would still try to determine the `count` of results in the queryset, so that doesn't fully help.